AnnService/inc/Core/Common/Dataset.h

```diff
@@ -60,39 +63,65 @@ namespace SPTAG
             void SetR(int R_) 
             {
                 if (R_ >= rows)
-                    dataIncremental.resize(((size_t)(R_ - rows)) * cols);
+                    incRows = R_ - rows;
                 else 
                 {
                     rows = R_;
-                    dataIncremental.clear();
+                    incRows = 0;
                 }
             }
-            inline int R() const { return (int)(rows + dataIncremental.size() / cols); }
+            inline int R() const { return rows + incRows; }
             inline int C() const { return cols; }
-            T* operator[](int index)
+            
+            inline const T* At(int index) const
             {
                 if (index >= rows) {
-                    return dataIncremental.data() + ((size_t)(index - rows)) * cols;
+                    int incIndex = index - rows;
+                    return incBlocks[incIndex / rowsInBlock] + ((size_t)(incIndex % rowsInBlock)) * cols;
                 }
                 return data + ((size_t)index) * cols;
             }
+
+            T* operator[](int index)
+            {
+                return (T*)At(index);
+            }
             
             const T* operator[](int index) const
             {
-                if (index >= rows) {
-                    return dataIncremental.data() + ((size_t)(index - rows)) * cols;
-                }
-                return data + ((size_t)index) * cols;
+                return At(index);
             }
 
             void AddBatch(const T* pData, int num)
             {
-                dataIncremental.insert(dataIncremental.end(), pData, pData + ((size_t)num) * cols);
+                int written = 0;
+                while (written < num) {
+                    int curBlockIdx = (incRows + written) / rowsInBlock;
+                    if (curBlockIdx >= incBlocks.size()) {
+                        incBlocks.push_back((T*)aligned_malloc(((size_t)rowsInBlock) * cols * sizeof(T), ALIGN));
```

Comment:
please notice when you push element into incBlocks, the data of incBlocks may be re-alloc too.